scripts/devaccess/devaccess_wrap.py

```diff
@@ -0,0 +1,455 @@
+
+    def parse_config(self):
+        ''' Load in config settings '''
+        self._config = yaml.load(open(DevGet.CONFIG_FILE, 'r'))
+
+        if not self._config.has_key('aclfile_path'):
+            self._config['aclfile_path'] = DevGet.ACL_FILE
+
+        if not self._config.has_key('logfile_path'):
+            self._config['logfile_path'] = DevGet.LOG_FILE
+
+        if not self._config.has_key('debug'):
+            self._config['debug'] = False
+
+    def setup_permissions(self):
+        ''' Read in users and roles.
+            Return list of allowed commands for user. '''
+
+        perm_dict = yaml.load(open(self._config['aclfile_path'], 'r'))
+        # create list of allowed commands for the user
+        for user in perm_dict['users']:
+            if user['username'] == self._user:
+                allowed_roles = user['roles']
+                # the 'ALL' group applies to everyone
+                allowed_roles.append('ALL')
+        logging.debug("user: %s roles: %s", self._user, str(allowed_roles))
+
+        commands = []
+        # get list of allowed commands for each role
+        for role in perm_dict['roles']:
+            if role['name'] in allowed_roles:
+                commands.extend(role['commands'])
+
+        logging.debug("user: %s commands: %s", self._user, str(commands))
+        return commands
+
+    @staticmethod
+    def whitelisted_command_list():
+        ''' Dict with key of whitelisted commmands mapped to their
+            actual implementation.
+            The commands should be in 'normalized' output style from OCCmd.
+        '''
+        command_dict = {}
+
+        ### commands for everyone (the 'ALL' group)
+        command_dict['oc get nodes -ndefault'] = WhitelistedCommands.oc_get_nodes
+        command_dict['oc get nodes -ndefault -ojson'] = WhitelistedCommands.oc_get_nodes
+        command_dict['rpm -qa'] = WhitelistedCommands.rpm_qa
+
+        return command_dict
+
+    def parse_args(self):
+        ''' Parse command line arguments passed in through the
+            SSH_ORIGINAL_COMMAND environment variable when READ_SSH is a
+            param.
+            Also handle when run manually. '''
+        args = None
+        user = None
+        read_ssh_env = False
+
+        # authorized_keys will force direct our command/argv to be
+        # 'devaccess_wrap READ_SSH <username>' with the original params stored
+        # in SSH_ORIGINAL_COMMAND
+        if "READ_SSH" in sys.argv:
+            read_ssh_env = True
+
+        if read_ssh_env:
+            cmd = os.environ.get("SSH_ORIGINAL_COMMAND", "")
+
+            # SSH_ORIGINAL_COMMAND will include the whole command
+            # as a continuous string
+            args = cmd
+
+            # The user's authorized_keys will force the command run to be:
+            # <path to devaccess_wrap> READ_SSH <username>
+            # Save the username for permission lookups later.
+            user = sys.argv[2]
```

Comment:
Two ways I can see to fake this.
1) User modifies .ssh/authorized_keys to replace the user argument. This implies that they have write access to .ssh/authorized_keys, so they can just remove the 'command=...' wrapper.
2) User 'bob' tries to run this wrapper script and sets the appropriate variables to get the script to read 'bob'-controlled argv. At this point, the script is running with the 'bob' permissions, so 'bob' wont be able to do anything that 'bob' couldn't already do.

I can't think of another way this could be attacked right now...